File: 3753. total-waviness-of-numbers-in-range-ii.java
class Solution {

    static class Node {
        long cnt;
        long wav;

        Node(long cnt, long wav) {
            this.cnt = cnt;
            this.wav = wav;
        }
    }

    String s;
    Node[][][][] dp = new Node[20][2][11][11];
    boolean[][][][] vis = new boolean[20][2][11][11];

    private Node dfs(int pos, int started, int last,
                     int secondLast, boolean tight) {

        if (pos == s.length()) {
            return new Node(1, 0);
        }

        if (!tight && vis[pos][started][last][secondLast]) {
            return dp[pos][started][last][secondLast];
        }

        int limit = tight ? s.charAt(pos) - '0' : 9;

        long totalCnt = 0;
        long totalWav = 0;

        for (int d = 0; d <= limit; d++) {
            boolean ntight = tight && (d == limit);

            if (started == 0 && d == 0) {
                Node nxt = dfs(pos + 1, 0, 10, 10, ntight);

                totalCnt += nxt.cnt;
                totalWav += nxt.wav;
            } else {
                long add = 0;

                if (started == 1 && secondLast != 10) {
                    if ((last > secondLast && last > d) ||
                        (last < secondLast && last < d)) {
                        add = 1;
                    }
                }

                int nSecondLast = (started == 1) ? last : 10;

                Node nxt = dfs(pos + 1, 1, d, nSecondLast, ntight);

                totalCnt += nxt.cnt;
                totalWav += nxt.wav + add * nxt.cnt;
            }
        }

        Node res = new Node(totalCnt, totalWav);

        if (!tight) {
            vis[pos][started][last][secondLast] = true;
            dp[pos][started][last][secondLast] = res;
        }

        return res;
    }

    private long solve(long n) {
        if (n < 0) return 0;

        s = Long.toString(n);

        vis = new boolean[20][2][11][11];
        dp = new Node[20][2][11][11];

        return dfs(0, 0, 10, 10, true).wav;
    }

    public long totalWaviness(long num1, long num2) {
        return solve(num2) - solve(num1 - 1);
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        long num1_1 = 120;
        long num2_1 = 130;
        long output1 = sol.totalWaviness(num1_1, num2_1);
        System.out.println("Example 1: num1=120, num2=130");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 3");
        System.out.println();
        
        // Example 2
        long num1_2 = 198;
        long num2_2 = 202;
        long output2 = sol.totalWaviness(num1_2, num2_2);
        System.out.println("Example 2: num1=198, num2=202");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 3");
        System.out.println();
        
        // Example 3
        long num1_3 = 4848;
        long num2_3 = 4848;
        long output3 = sol.totalWaviness(num1_3, num2_3);
        System.out.println("Example 3: num1=4848, num2=4848");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 2");
    }
}

File: 3753. total-waviness-of-numbers-in-range-ii.py
class Solution:
    def totalWaviness(self, num1: int, num2: int) -> int:

        from functools import lru_cache

        def solve(n: int) -> int:
            if n < 0:
                return 0

            s = str(n)

            @lru_cache(None)
            def dfs(pos, started, last, second_last, tight):
                if pos == len(s):
                    return (1, 0)

                limit = int(s[pos]) if tight else 9

                total_cnt = 0
                total_wav = 0

                for d in range(limit + 1):
                    ntight = tight and d == limit

                    if not started and d == 0:
                        cnt, wav = dfs(
                            pos + 1,
                            False,
                            10,
                            10,
                            ntight
                        )

                        total_cnt += cnt
                        total_wav += wav

                    else:
                        add = 0

                        if started and second_last != 10:
                            if (
                                (last > second_last and last > d)
                                or
                                (last < second_last and last < d)
                            ):
                                add = 1

                        n_second_last = last if started else 10

                        cnt, wav = dfs(
                            pos + 1,
                            True,
                            d,
                            n_second_last,
                            ntight
                        )

                        total_cnt += cnt
                        total_wav += wav + add * cnt

                return (total_cnt, total_wav)

            return dfs(0, False, 10, 10, True)[1]

        return solve(num2) - solve(num1 - 1)

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    num1_1 = 120
    num2_1 = 130
    output1 = sol.totalWaviness(num1_1, num2_1)
    print(f"Example 1: num1={num1_1}, num2={num2_1}")
    print(f"Output: {output1}")
    print(f"Expected: 3")
    print()
    
    # Example 2
    num1_2 = 198
    num2_2 = 202
    output2 = sol.totalWaviness(num1_2, num2_2)
    print(f"Example 2: num1={num1_2}, num2={num2_2}")
    print(f"Output: {output2}")
    print(f"Expected: 3")
    print()
    
    # Example 3
    num1_3 = 4848
    num2_3 = 4848
    output3 = sol.totalWaviness(num1_3, num2_3)
    print(f"Example 3: num1={num1_3}, num2={num2_3}")
    print(f"Output: {output3}")
    print(f"Expected: 2")